### PR TITLE
DEPR: dropping nuisance columns in DataFrameGroupby apply, agg, transform

### LINKED ISSUE #21664: Dropping nuisance columns in groupby is a nuisance
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from decimal import Decimal
df = pd.DataFrame({'id': [1], 'x': [1], 'y': [Decimal(1)]})
df.groupby('id')[['x', 'y']].sum()

#     x
# id   
# 1   1
```
#### Problem description

I unknowingly encountered the feature described [here](http://pandas.pydata.org/pandas-docs/stable/groupby.html#automatic-exclusion-of-nuisance-columns) when running the above code. While I see how this can be a useful feature, it's a nuisance not knowing that it happened and that I can't disable it. I feel that in the case of doing a groupby on explicitly selected columns `groupby(...)[COLS]`, it should not drop any columns and let whatever errors that occur raise. I also think that a warning could be added and/or an option to disable the feature.

#### Expected Output
```
#     x  y
# id   
# 1   1  1
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.10.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None
pandas: 0.22.0
pytest: 3.2.1
pip: 9.0.1
setuptools: 36.4.0
Cython: 0.28.1
numpy: 1.13.1
scipy: None
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #21664
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Discussed on this week's call.

### CODE REVIEW

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -647,6 +647,7 @@ Deprecations
 - Deprecated setting :attr:`Categorical._codes`, create a new :class:`Categorical` with the desired codes instead (:issue:`40606`)
 - Deprecated behavior of :meth:`DatetimeIndex.union` with mixed timezones; in a future version both will be cast to UTC instead of object dtype (:issue:`39328`)
 - Deprecated using ``usecols`` with out of bounds indices for ``read_csv`` with ``engine="c"`` (:issue:`25623`)
+- Deprecated ignoring "nuisance columns" in :meth:`DataFrameGroupBy.agg`, :meth:`DataFrameGroupBy.apply`, and :meth:`DataFrameGroupBy.transform`; select valid columns before calling the method (:issue:`41475`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:650]: can you make this a sub-section (ok to lump this in with the DataFrame reductions PR sub-section as well), e.g. just a hey we are now deprecating both of these simliar things, with an example.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.0.rst (hunk 1 above)
Comment (jreback) [reply doc/source/whatsnew/v1.3.0.rst:650]: e.g. #41480

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.3.0.rst]: updated + green

**Review by jreback — APPROVED** [review]

**Comment by rhshadrach** [comment]
In `groupby._transform`, I think we can always use `_wrap_transform_fast_result` in the case of a string aggregator instead of falling back to `_transform_general` for certain cases. The only thing stopping this now is it removes the warnings added here. Digging into it, I'm finding some inconsistencies:

```
df = DataFrame(
    {
        "A": ["foo", "bar", "foo", "bar", "foo", "bar", "foo", "foo"],
        "B": ["one", "one", "two", "three", "two", "two", "one", "three"],
        "C": np.random.randn(8),
        "D": np.random.randn(8),
    }
)
df.groupby("A").mean()
df.groupby("A").agg("mean")
df.groupby("A").agg(np.mean)
df.groupby("A").transform("mean")  # <--- warns
df.groupby("A").transform(np.mean)  # <--- warns
df.groupby("A").apply("mean")
df.groupby("A").apply(np.mean)  # <--- warns
```

Should all, just the `np.mean`, or none of these warn? From the [whatsnew note](https://pandas.pydata.org/docs/whatsnew/v1.3.0.html#deprecated-dropping-nuisance-columns-in-dataframe-reductions-and-dataframegroupby-operations), it seemed to me that all of these should warn.

**Comment by jbrockmendel** [comment]
> Should all, just the np.mean, or none of these warn? From the [whatsnew note](https://pandas.pydata.org/docs/whatsnew/v1.3.0.html#deprecated-dropping-nuisance-columns-in-dataframe-reductions-and-dataframegroupby-operations), it seemed to me that all of these should warn.

I think you're right.  IRC getting consistency across all these cases was something we punted on.

**Comment by rhshadrach** [comment]
Thanks, I'll take a deeper look into emitting the warning in the right places.

**Comment by rhshadrach** [comment]
I missed that the default of `numeric_only` is still currently True, so many of the above cases are currently correct. With #46072, all of the `.mean()` and `"mean"` forms will warn. The `np.mean` should warn in all cases, but doesn't for `agg`. I plan to be putting up a PR for this.

**Pull request merged**